Design proposal for configurable kubecontext (#2384)
GoogleContainerTools/skaffold: docs/design_proposals/configurable-kubecontext.md

# Configurable kubecontext

* Author(s): Cornelius Weig (@corneliusweig)

> * Design Shepherd: tbd

Reviewer: ```suggestion
* Design Shepherd: Balint Pato (@balopat)
```

There also was an attempt to add a configuration option to `skaffold.yaml` (Support for overriding kubectl context during deployment #1540).

The goal of this document is to create an agreement on what options should be supported and identify edge cases.

## Design

There are four places where kubecontext activation can be added:
<table>
    <thead>
        <th>Precedence</th> <th>Kind</th> <th>Details</th>
    </thead>
    <tbody>
        <tr>
            <td>1. (highest)</td>
            <td>CLI option</td>
            <td>
              The kubernetes standard to set the kubecontext is <code>--context</code>.
              However, in Skaffold this term is so overloaded that it should more precisely be named <code>--kube-context</code>.
              This flag is necessary for IDE integration.
            </td>
        </tr>
        <tr>
            <td>2.</td>
            <td>Env variable</td>
            <td>
              <code>SKAFFOLD_KUBE_CONTEXT</code>, similar to other Skaffold flags.
            </td>
        </tr>
        <tr>
            <td>3.</td>
            <td><code>skaffold.yaml</code></td>
            <td>

>               Json-path <code>deploy.kubeContext</code>.

Reviewer: ```suggestion
              Json-path <code>[profile.]deploy.kubeContext</code>.
```

Author: My current feeling is that we should have the option also in the main pipeline. Although this may not be the recommended way to use this feature, I have a hunch that this will be requested nevertheless.

Reviewer: agreed, `[profile.]` meant to signify main pipeline or optionally in a profile...maybe there is a better way to express that :)

            <td>
              This should give users the possibility to define a default context globally or per project.
              This variant is not shareable.
            </td>
        </tr>
    </tbody>
</table>

---

Beside the kubecontext, also the namespace needs to be specified.
It should also be possible to specify in various places:

1. `--namespace` CLI flag
2. `SKAFFOLD_NAMESPACE` env var
3. In `skaffold.yaml` at json-path `deploy.namespace`
4. globally or per project in the Skaffold config

### Detailed discussion
#### Option in `skaffold.yaml`
A configuration option in `skaffold.yaml` has the advantage of being most discoverable:
it is in the place where users configure all aspects of their pipeline.
In addition, it allows to define the kubecontext per Skaffold profile.

However, it also has questionable implications:

- `skaffold.yaml` is meant to be shared, but kubecontext names vary across users.
  Sharing therefore makes only sense in a corporate setting where context names are the same across many users.
  There is however a risk of abuse in settings where sharing the context name does not make sense, for example in open source projects.

> - Due to profile activation by Skaffold profiles, there can be circular problems.

Reviewer: ```suggestion
- Due to profile activation by Skaffold profiles, there can be confusing/surprising situations.
```

This is not really a circular problem as we don't change the "current context".

>   For example, the current context is `A` and activates some profile.

Reviewer: ```suggestion
  For example, the current context is `minikube` and activates some profile.
```

>   This profile activates a different kubecontext `B`.

Reviewer: ```suggestion
  This profile is deploying to a different kubecontext `B`. This can be surprising to the user.
```

>   A solution could be to forbid specifying a kubecontext in a profile if it is activated by a kubecontext (and validate that).

Reviewer: This should be the recommended solution. 
We should mention that this check should on the "effective profile"  that is the result of merging all the activated profiles.

Author: I'm a bit puzzled about the effective profile. If we forbid to override the kubecontext twice by two profiles, then we are deviating from the normal behavior of how profiles are usually applied. Usually profiles are applied one after the other, and last one wins. So why should that be different for the kubecontext?

I think a static check where a profile with activation by kubecontext must not override the kubecontext is easy to understand and fine. However, as soon as the logic becomes dynamic this quickly becomes rather intransparent.

Reviewer: Hm...but then 

given: current context = minikube 
activated profiles: 

P1 <- envvar1
P2 <- cur_ctx=minikube  

and P1.deploy.kubecontext = gke_foobar

I'll end up with the minikube specific profile that deploys to gke...

Author: I would say

given: current context = minikube

P1 <- envvar1,   changes the kubeContext to `gke_foobar`
P2 _no longer applies_, because kubeContext is not `minikube` anymore

---

I just checked the `skaffold.yaml` and the profile activation by kube-context is simply called `kubeContext`. It does not necessarily mean that it only considers the current context as defined in the kubeconfig. That would even be very misleading: when somebody changes the kube-context on the CLI with `skaffold --kube-context gke_foobar`, but the kubeconfig says `minikube`, I would certainly not expect that Skaffold activates a minikube profile.

Author: I think profile activation is confusing, and one can construct pretty contrived examples:

given: current context = `minikube`
P1 <- envvar, changes kubeContext to `gke_foobar`
P2 <- activated by kubeContext `gke_foobar`
P3 <- activated by envvar, changes kubeContext to `minikube`

In that sequence, the gke-specific profile would be activated on minikube. :zap: 

---

This example suggests that a dynamic check is necessary in any case. I think the validation should look out for profiles that were activated by a kube-context which is no longer active in the effective `skaffold.yaml` (Pardon me, if that is what you were trying to tell me all along).

Author: Never mind my previous comments. I should have read the code first.

So profile activation and profile application are separate, meaning that all profile activations use the same current kubecontext. Iow there can be no transitive kubecontext activation like
```
P1 activates minikube
P2 is activated by minikube
...and so on
```
A reasonable validation is therefore:
- check if any profile is activated by the current kubecontext before profile activation
- after profile application, check if the effective kubecontext `skaffold.yaml` has changed.
- if yes, and there was a context-activated profile, return error
- otherwise, everything is fine.
(Pardon me, if this is what you wanted to tell me all along..)
This check also makes any static check of the `skaffold.yaml` obsolete.

The case where subsequent profiles repeatedly switch the kubecontext is probably fine, as profile activation follows a last-one-wins policy which should also apply to kubecontext.

Reviewer: This sounds good, yes, this is exactly what I meant :)

A natural place for the config in `skaffold.yaml` is in `latest.DeployConfig`, resulting in a json path `deploy.kubeContext`.

#### Option in global Skaffold config
Specifying a default kubecontext globally is straightforward. For example, via new config option
```yaml
global:
  default-context: docker-for-desktop
```

On the other hand, building a relation between projects and kubecontext needs to solve two questions:

1. How to identify projects
2. How to save the project/kubecontext relation

##### How to identify projects

There are at least two possibilities:

- Identify projects by their absolute host path.
  This is guaranteed to be unique, but may break if a user moves his project to another location.

> - Identify projects by a new `metadata.name` entry in `skaffold.yaml` (see also #2200).

Reviewer: ```suggestion
- Identify projects by a new `metadata.name` entry in `skaffold.yaml` (see also [#2200](https://github.com/GoogleContainerTools/skaffold/issues/2200)).
```

  This has the drawback of being potentially not unique, so that users accidentally pin the kubecontext for more projects than intended.
  On the other hand, this is the standard approach taken by kubernetes resources.
- Identify project by their initial commit.
  This variant is stable against relocations.
  It is also unique unless a user forks a project and wants to define different kubecontexts for each fork.
  Compared to the other alternatives, it is rather slow.

**\<What option has the best tradeoffs?\>**

Resolution: We will go ahead with the `metadata.name` approach. As the name may not be unique, this requires special documentation.

##### How to save project/kubecontext relations

Currently, the Skaffold config uses the kubecontext as identifier.

There are two possibilities to add the relation:

- Reverse the mapping project/kubecontext and save as list under `kube-context` entries:
  ```yaml
  kubecontexts:
  - kube-context: my-context

>     projects:

Author: @balopat Please correct me if I'm wrong, but my impression was that this variant was the preferred one.

Also, we should carefully think about how to name the new field, as the global config is not versioned and field renaming will be much harder. @tejal29 confused it with GCB projects, and this may happen to other users too. We could make it clearer that this ties the kubecontext to a specific named `skaffold.yaml`, and call the field `skaffoldConfigs` instead.

Reviewer: +1 on `skaffoldConfigs`

    - /home/user/projects/skaffold-project
    - project-name
    - d76524a31c58657b06404f4442e1e64949f4d762
  ```
  The drawback here is that the data structure does not forbid conflicting entries, such as this:
  ```yaml
  kubecontexts:
  - kube-context: context1
    projects:
    - my-project
  - kube-context: context2
    projects:
    - my-project
  ```
- Add a new top-level entry in Skaffold config:
  ```yaml
  global: {}
  kubecontexts: []
  projects:
    my-project: # project name as the key
      kube-context: context1
  ```
  This option will be more complex to implement wrt `skaffold config`.

**\<What Skaffold config structure has the best tradeoffs?\>**

Resolution: __Not Yet Resolved__

### Open Issues/Questions

**\<Should there be a config option in `skaffold.yaml`?\>**

> Resolution: Yes.

Author: I think the overall agreement was that we do want to have the option in `skaffold.yaml`

Reviewer: Yes, confirmed!

- Identify project by their initial commit.
  This variant is stable against relocations.
  It is also unique unless a user forks a project and wants to define different kubecontexts for each fork.
  Compared to the other alternatives, it is rather slow.

**\<What option has the best tradeoffs?\>**

Resolution: We will go ahead with the `metadata.name` approach. As the name may not be unique, this requires special documentation.

##### How to save project/kubecontext relations

Currently, the Skaffold config uses the kubecontext as identifier.

There are two possibilities to add the relation:

- Reverse the mapping project/kubecontext and save as list under `kube-context` entries:
  ```yaml
  kubecontexts:
  - kube-context: my-context
    skaffoldConfigs:
    - config-name
  ```
  The drawback here is that the data structure does not forbid conflicting entries, such as this:
  ```yaml
  kubecontexts:
  - kube-context: context1
    skaffoldConfigs:
    - my-project
  - kube-context: context2
    skaffoldConfigs:
    - my-project
  ```
- Add a new top-level entry in Skaffold config:
  ```yaml
  global: {}
  kubecontexts: []

>   projects:

Author: Should be 
```yaml
skaffoldConfigs:
  my-project: my-context
  my-other-project: my-other-context
```

    my-project: # project name as the key
      kube-context: context1
  ```
  This option will be more complex to implement wrt `skaffold config`.

### Open Issues/Questions

**\<What Skaffold config structure has the best tradeoffs?\>**

> Resolution: __Not Yet Resolved__

Author: @balopat Let's resolve this last open question, so that this design proposal can be merged.

My feeling is that option 2 (map-like structure) is overall better suited to store project-to-context mappings. Besides, option 1 (`skaffoldConfigs` list under kubecontext entry) requires additional special cases for the `skaffold config set/unset` commands. So I'm more inclined towards the second option:
```yaml
skaffoldConfigs:
  my-project: my-context
  my-other-project: my-other-context
```
WDYT?

Author: I just found another advantage of option 2: It allows a much more seamless implementation of the wildcard global default context like so:
```yaml
skaffoldConfigs:
  '*': default-context
  my-project: my-context
```

Reviewer: I'm okay with this direction!
